### PR TITLE
Add simple monitoring ldap dependencies

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: cmd/keymasterd/dependency_monitor.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import (
+	"crypto/x509"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/Symantec/keymaster/lib/authutil"
+
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (rgooch) [inline cmd/keymasterd/dependency_monitor.go]: Please also add tricorder metrics.

File: cmd/keymasterd/dependency_monitor.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+const timeoutSecs = 5
+
+func init() {
+	prometheus.MustRegister(dependencyLatency)
+	prometheus.MustRegister(dependencyLastSuccessSecondsGauge)
+}
+
+func checkLDAPURLs(ldapURLs string, name string, rootCAs *x509.CertPool) error {
```
Comment (rgooch) [inline cmd/keymasterd/dependency_monitor.go]: There should probably be a timeout for the whole function. If you have a lot of URLs to try and each has a 5 second timeout, you probably want to declare failure before they all time out.

File: cmd/keymasterd/main.go [hunk 3]
```diff
@@ -147,6 +147,7 @@ type RuntimeState struct {
 const redirectPath = "/auth/oauth2/callback"
 const secsBetweenCleanup = 30
 const maxAgeU2FVerifySeconds = 30
+const secsBetweenDependencyChecks = 60
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: Please make this configurable. Either a command-line flag or config file option.

File: cmd/keymasterd/dependency_monitor.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			Name: "keymaster_dependency_durations_since_last_success_seconds",
+			Help: "Seconds since last update",
+		},
+		[]string{"type", "name"},
+	)
+	lastSuccessLDAPPasswordTime time.Time
+	lastSuccessLDAPUserInfoTime time.Time
+)
+
+const timeoutSecs = 5
```
Comment (rgooch) [inline cmd/keymasterd/dependency_monitor.go]: Please make this configurable.

File: lib/authutil/authutil.go [hunk 5]
```diff
@@ -137,6 +137,18 @@ func getLDAPConnection(u url.URL, timeoutSecs uint, rootCAs *x509.CertPool) (*ld
...
+func CheckLDAPConnection(u url.URL, timeoutSecs uint, rootCAs *x509.CertPool) error {
+	conn, _, err := getLDAPConnection(u, timeoutSecs, rootCAs)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	timeout := time.Duration(time.Duration(timeoutSecs) * time.Second)
+	conn.SetTimeout(timeout)
+	conn.Start()
+	return nil
```
Comment (rgooch) [inline lib/authutil/authutil.go]: This always returns OK? Regardless of timeouts?

File: cmd/keymasterd/dependency_monitor.go (hunk 1 above)
Comment (cviecco) [reply cmd/keymasterd/dependency_monitor.go]: done

File: cmd/keymasterd/dependency_monitor.go (hunk 2 above)
Comment (cviecco) [reply cmd/keymasterd/dependency_monitor.go]: usually there are Not a lot of urls, and it is fine if they reach the timeout

File: cmd/keymasterd/main.go (hunk 3 above)
Comment (cviecco) [reply cmd/keymasterd/main.go]: done

File: cmd/keymasterd/dependency_monitor.go (hunk 4 above)
Comment (cviecco) [reply cmd/keymasterd/dependency_monitor.go]: Yagni (I am pulling an apple here)

File: lib/authutil/authutil.go (hunk 5 above)
Comment (cviecco) [reply lib/authutil/authutil.go]: yes

**Pull request merged**